### PR TITLE
Revert "KAFKA-16154"  and mark MV 3.8-IV0 as the latest production

### PR DESCRIPTION
**Pull request by jlprat**
This reverts commit 55a6d30c
As described in https://github.com/apache/kafka/pull/16347#issuecomment-2176593301

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jlprat** [comment]
@cmccabe If you have time, this should be the revert of commit https://github.com/apache/kafka/commit/55a6d30ccbe971f4d2e99aeb3b1a773ffe5792a2 on the 3.8 branch.
A generated revert was not possible as there were merge conflicts.

File: core/src/main/scala/kafka/log/UnifiedLog.scala
Comment (jlprat) [inline core/src/main/scala/kafka/log/UnifiedLog.scala:1335]: These and the 8 lines prior were the ones with a conflict when reverting.

**Comment by jlprat** [comment]
Build failure contained:
> 14:54:56  hudson.remoting.Channel$OrderlyShutdown: Command Close created at ...

I triggered another build

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:239]: 1. 3.8_IV0 is no longer controlling ELR. We need to change the following comment according to https://github.com/apache/kafka/pull/16347/files#diff-abe728e9704c4c6709c4c3f6fd3908d51219cda55791ac7f103e6042c1d6c298.
```
    // Add ELR related supports (KIP-966).
    IBP_3_8_IV0(20, "3.8", "IV0", true),
```
Also, quite a few tests (e.g. [PartitionChangeBuilderTest.java](https://github.com/apache/kafka/pull/16347/files#diff-4fd9ae1562f036f713a0501fa89ebf4f68e8bfc25c046504f885df94ff48a371)) still reference 3.8_IV0 with the assumption that it's associated with ELR. @cmccabe: Should we bring IBP_3_9_IV0, the new MV associated with ELR to the 3.8 branch or change all those tests related to ELR?

2. There are a few tests referring to IBP_3_7_IV4, the previous latest production version. They all need to be changed to the new latest production version IBP_3_8_IV0.
 
[ApiVersionsRequestTest.scala](https://github.com/apache/kafka/pull/16347/files#diff-661b7ee4b762dcace5958cf95f7e31762fccf4dc3df571b0ca3ed80036e8cbc7)

[FeatureCommandTest.java](https://github.com/apache/kafka/pull/16347/files#diff-ee0495307709e369188ef829cde8a0be4df32f9e4a0dc932d530d4052efbcc53)

3. We need to add a comment above the following line that 3.8.0 is the latest production version in 3.8 line.

https://github.com/apache/kafka/blob/3.8/server-common/src/test/java/org/apache/kafka/server/common/MetadataVersionTest.java#L187

**Review by junrao — COMMENTED** [review]
@jlprat : Thanks for the PR. Added a few comments.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:239]: Applied feedback, except for "Should we bring IBP_3_9_IV0, the new MV associated with ELR to the 3.8 branch or change all those tests related to ELR?". I'm waiting for @cmccabe's feedback on this one.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:205]: I believe this was an oversight while doing the revert. 3.8-IV0 should have no metadata changes

File: server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java
Comment (jlprat) [inline server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java]: @jolshan I need help understanding this test setup. As far as I could understand, this `TEST_2` feature was meant to be released in a future release, given we 3.8 is the current one, I moved it to 4.0. But I think this is not a robust test. Probably I'm missing something else.

File: tools/src/test/java/org/apache/kafka/tools/FeatureCommandTest.java
Comment (jlprat) [inline tools/src/test/java/org/apache/kafka/tools/FeatureCommandTest.java]: This needed to change as well as we changed the parameter of the test to 3.8-IV0

File: server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java
Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java]: I plan to update this in a future PR. Sorry it is not easy to understand now. If you can push it to 4.0, that would be good for now.

**Comment by jlprat** [comment]
Build failed with an exception on "streams:upgrade-system-tests-34:test". Rerunning the build

**Comment by junrao** [comment]
@jlprat : The PR now has conflicts after https://github.com/apache/kafka/pull/16478. Could you rebase? Thanks.

**Comment by jlprat** [comment]
@junrao Rebased
@jolshan I would highly appreciate your review for "marking IBP_3_8_V0 as latest production". There are couple of changes I needed to do to make the tests pass.

**Comment by jlprat** [comment]
Something seems wrong after rebasing. Build hangs and the job gets terminated.
Apparently `ReplicationQuotasTest > shouldBootstrapTwoBrokersWithLeaderThrottle(String) > "shouldBootstrapTwoBrokersWithLeaderThrottle(String).quorum=kraft"` never terminates.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Hmm, we removed GroupVersion in 3.8 https://github.com/apache/kafka/pull/16478. Why are we adding it back? The hanging ReplicationQuotasTest is related to this since registerBroker() needs to set MaxSupportedVersion to the latest MV, which is 4.0 now. See a similar change in trunk https://github.com/apache/kafka/pull/16347/files#diff-4f8200cb972ea4c0d08bcfc45e9e4e6295c37a4414b9acd59cc4bb248bc25a1c

**Review by junrao — COMMENTED** [review]
@jlprat : Thanks for the updated PR. Left a comment.

**Comment by junrao** [comment]
Since this PR also marks MV 3.8-IV0 as the latest production, it would be useful to update the description of the PR accordingly.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Good pointer, I think this might be the problem. This was probably a wrong conflict resolution during rebasing.
Why this existed and was used in this PR is because due to the test setup for FeaturesTest.
`TEST_2` needs to be configured to 4.0 like this https://github.com/apache/kafka/pull/16400/files#diff-10fd2b01213aa0cc1f879b06eb7eaae357e3dbcf8a270660b0eb4fcc78e7a5a1R27
Otherwise these tests fail:
```
:server-common:test > Gradle Test Executor 4 > FeaturesTest > testDefaultTestVersion(MetadataVersion) > "testDefaultTestVersion(MetadataVersion).metadataVersion=3.8-IV0" FAILED
    org.opentest4j.AssertionFailedError: expected: <1> but was: <2>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at app//org.junit.jupiter.api.AssertEquals.failNotEqual(AssertEquals.java:197)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:134)
        at app//org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:129)
        at app//org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:329)
        at app//org.apache.kafka.server.common.FeaturesTest.testDefaultTestVersion(FeaturesTest.java:120)
```

```
Gradle Test Run :server-common:test > Gradle Test Executor 4 > FeaturesTest > testUnstableTestVersion() FAILED
    org.opentest4j.AssertionFailedError: Expected java.lang.IllegalArgumentException to be thrown, but nothing was thrown.
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:152)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:73)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
        at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3115)
        at app//org.apache.kafka.server.common.FeaturesTest.testUnstableTestVersion(FeaturesTest.java:125)
```

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I'll remove it, and try to find out how to fix the tests that seem to expect MV 3.8-IV0 to not be the production one.

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I think if you need an unstable version, you can just use 4.0 (or 3.9)

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: @jolshan Isn't this what is being done in this PR? So if I add the 4.0, the other tests work, but then `ReplicationQuotasTest` hangs.
This area is not one that I'm really familiar with.

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Hmm. I'm not sure I follow how the tests are related. I can pull your branch if that helps.

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: @jolshan https://github.com/apache/kafka/pull/16400/files#diff-10fd2b01213aa0cc1f879b06eb7eaae357e3dbcf8a270660b0eb4fcc78e7a5a1R27 needs to be configured as 4.0 (or 3.9 for that matter) so the `FeaturesTest` works. But if I add 4.0 in `MetadaVersion` then `ReplicationQuotasTest.shouldBootstrapTwoBrokersWithLeaderThrottle` hangs.

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I think I found a way. I'll push shortly

File: core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala
Comment (jlprat) [inline core/src/test/scala/unit/kafka/server/ReplicationQuotasTest.scala:302]: This makes the test pass as it disregards the latest version as per https://github.com/apache/kafka/pull/16347

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java
Comment (jlprat) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:199]: As per https://github.com/apache/kafka/pull/16347

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Taking changes from https://github.com/apache/kafka/pull/16347

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jolshan) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: My understanding from trunk is that we removed the reference to KIP-966. metadataChanged should be false.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: It's ok to introduce an unstable 4.0 MV for the TestFeature. We just need to (1) make sure the comment associated with the new MV is correct; (2) change all tests that depend on the latest testing MV. In https://github.com/apache/kafka/pull/16347, we tried to change all relevant tests to use MetadataVersion.latestTesting() so that we don't have to manually change the MV when new unstable MV is added in the future. You can search for all tests that have changed from MetadataVersion.IBP_4_0_IV0 to MetadataVersion.latestTesting() in https://github.com/apache/kafka/pull/16347 and apply them here.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: See https://github.com/apache/kafka/blob/206d0f809ad8438e9d89b7f731a8b995f952f0c0/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L214

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I had this set to false in one of the commits, but then I reverted it back to true, because of rebasing.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I did this now for all the changes I belive

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: 1. The following comment needs to be moved to IBP_3_9_IV0.

`// Add ELR related supports (KIP-966).`

2. We need to change `isElrSupported()` to use IBP_3_9_IV0.
3. We need to bring all tests referencing IBP_3_9_IV1 for ELR in https://github.com/apache/kafka/pull/16347 here.
4. We need to bring in the changes in the following tests in https://github.com/apache/kafka/pull/16347 for testing latestTesting MV.
ClusterTestExtensionsTest.java
ZkMigrationIntegrationTest.scala
ApiVersionsRequestTest.scala
5. We need to add IBP_3_9_IV0 to all tests in MetadataVersionTest.java where 4.0-IV0 is used in trunk.

**Review by junrao — COMMENTED** [review]
@jlprat : Thanks for the updated PR. Left one more comment.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Thanks @junrao
Would we need to add IBP_3_9_IV1 as well? Or just use the tests for IBP_3_9_IV1 using the IBP_3_9_IV0 enum value?

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: Just use the IBP_3_9_IV0 enum value for tests using IBP_3_9_IV1 in trunk since BP_3_9_IV0 is associated with ELR in 3.8.

File: core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala
Comment (jlprat) [inline core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala]: `ClusterTest` in trunk has `IBP_3_9_IV0` as default value instead of `IBP_3_8_IV0`. This took me a while to detect as I didn't understand why this test was failing.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: I think I brought all necessary changes now

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:205]: nit: I think we should remove the ELR comment for 3.8

Comment (jlprat) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:205]: done

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java
Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java]: ELR changed metadata records.

**Review by jolshan — APPROVED** [review]
lgtm as long as the tests look good.

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java
Comment (junrao) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java]: `testUncleanShutdownBroker()` is based on ELR. So we need to change the 4 instances of IBP_3_8_IV0 in it to IBP_3_9_IV0. See https://github.com/apache/kafka/pull/16347

File: core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala
Comment (junrao) [reply core/src/test/scala/unit/kafka/server/ApiVersionsRequestTest.scala]: Actually, ClusterTest.metadataVersion() should default to latest testing MV, which is IBP_3_9_IV0. Could we change that and remove setting metadataVersion in this test and the one above?

**Review by junrao — COMMENTED** [review]
@jlprat : Thanks for the updated PR. A few more comments.

**Comment by cmccabe** [comment]
To respond to some of the questions above, overall,

- 3.8-IV0 in the 3.8 branch needs to be correct and completely in line with trunk

- 3.9-IV0 and 3.9-IV1 in the 3.8 branch can be whatever (it doesn't matter, since they are not usable / stable in 3.8)

- As you've probably noticed, the tests which use `@ClusterTest` often need to be manually updated when the latest MV changes. I do think this is a problem (with `@ClusterTest`). I don't have any quick solution right now other than doing the edits. This isn't an issue with the tests using TestKit directly, since they can just set `MetadataVersion.latestTesting`. Maybe we need to rethink the `@ClusterTest` stuff.

Let me know if you need me specifically to review

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java
Comment (jlprat) [reply metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java]: Indeed, this test was failing. I missed this change.

**Comment by jlprat** [comment]
Thanks for the review @junrao. Feedback is now applied

**Comment by jlprat** [comment]
Failing tests shouldn't be related to the changes (all tests passed locally)

**Review by junrao — APPROVED** [review]
@jlprat : Thanks for the updated PR. LGTM

**Review by jolshan — APPROVED** [review]
Reapproving with the latest changes! Thanks Josep.

File: metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java
Comment (junrao) [inline metadata/src/test/java/org/apache/kafka/controller/QuorumControllerTest.java:720]: Note that ideally, we should use MetadataVersion.latestTesting() here. However, explicitly using IBP_3_9_IV0 works too since it's the latest MV.

**Comment by jlprat** [comment]
Merged! Thanks for the reviews!

**Pull request merged**
